org.eclipse.scanning.connector.epicsv3/src/org/eclipse/scanning/connector/epicsv3/EpicsV3DynamicDatasetConnector.java

```diff
@@ -518,39 +505,62 @@ public void monitorChanged(MonitorEvent arg0) {
 							}
 							lastSystemTime = timeNow;
 						}
-						
-					} else if (channelName.equalsIgnoreCase(currentWidthPV)) {
-						DBR dbr = arg0.getDBR();
+					} else if (channelName.equalsIgnoreCase(dim0PV)) {
 						DBR_Int dbri = (DBR_Int)dbr;
-						if (width != dbri.getIntValue()[0]) {
-							width = dbri.getIntValue()[0];
-							dataChannelMonitor.removeMonitorListener(dataChannelMonitorListener);
-							int dataSize = height * width;
-							if (numDimensions == 3) {
-								dataSize *= rgbChannels;
-							}
-							dataChannelMonitor = ec.setMonitor(dataChannel, dataChannelMonitorListener, dataSize);
+						int value = dbri.getIntValue()[0];
+						if (dim0 != value) {
+							dim0 = value;
+							logger.debug("New value {} for {}", value, dim0PV);
+							updateDataChannelMonitor();
 						}
-					} else if (channelName.equalsIgnoreCase(currentHeightPV)) {
-						DBR dbr = arg0.getDBR();
+					} else if (channelName.equalsIgnoreCase(dim1PV)) {
 						DBR_Int dbri = (DBR_Int)dbr;
-						if (height != dbri.getIntValue()[0]) {
-							height = dbri.getIntValue()[0];
-							dataChannelMonitor.removeMonitorListener(dataChannelMonitorListener);
-							int dataSize = height * width;
-							if (numDimensions == 3) {
-								dataSize *= rgbChannels;
-							}
-							dataChannelMonitor = ec.setMonitor(dataChannel, dataChannelMonitorListener, dataSize);
+						int value = dbri.getIntValue()[0];
+						if (dim1 != value) {
+							dim1 = value;
+							logger.debug("New value {} for {}", value, dim1PV);
+							updateDataChannelMonitor();
+						}
+					} else if (channelName.equalsIgnoreCase(dim2PV)) {
+						DBR_Int dbri = (DBR_Int)dbr;
+						int value = dbri.getIntValue()[0];
+						if (dim2 != value) {
+							dim2 = value;
+							logger.debug("New value {} for {}", value, dim2PV);
+							updateDataChannelMonitor();
+						}
+					} else if (channelName.equalsIgnoreCase(numDimensionsPV)) {
+						DBR_Int dbri = (DBR_Int)dbr;
+						int value = dbri.getIntValue()[0];
+						if (numDimensions != value) {
+							numDimensions = value;
+							logger.debug("New value {} from {}", value, numDimensionsPV);
+							updateDataChannelMonitor();
 						}
+					} else if (channelName.equalsIgnoreCase(colourModePV)) {
+						DBR_String dbrs = (DBR_String)dbr;
+						String value = dbrs.getStringValue()[0];
+						if (colourMode != value) {
+							colourMode = value;
+							logger.debug("New value '{} 'from {}", value, colourModePV);
+							updateDataChannelMonitor();
+						}
+					} else {
+						logger.debug("New value from {}, ignoring: {}", channelName, dbr);
 					}
 				}
-				
 			} catch (Exception e) {
 				logger.error(e.getMessage());
 				e.printStackTrace();
 			}
 		}
-		
+	}
+
+	private void updateDataChannelMonitor() throws CAException, InterruptedException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Move this method into "EpicsMonitorListener". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS3398)